### PR TITLE
Add Example Tests

### PR DESCRIPTION
**Pull request by riusricardo**
- ERC20 tests

### CODE REVIEW

**Comment by athei** [comment]
You are not using `run_test`. Should you be using it or is this the new way (without it).

**Comment by Robbepop** [comment]
> You are not using `run_test`. Should you be using it or is this the new way (without it).

Not using `run_test` is actually wrong and I am still trying to find out what is broken with it atm. In the end what we want is a proc macro `#[ink(test)]` that handles all test-cases for you.

**Comment by athei** [comment]
That would be awesome. Getting rid of that indentation level.

**Comment by Robbepop** [comment]
> That would be awesome. Getting rid of that indentation level.

https://github.com/paritytech/ink/issues/335

**Comment by riusricardo** [comment]
I agree with you @athei and @Robbepop, I'm aware of this. This is why I left it as draft PR.
I did it the same as Alex and got into the bug. I'm just waiting for this bug to be fixed.

**Comment by Robbepop** [comment]
@riusricardo What is the state of this PR?

File: examples/erc20/src/lib.rs [hunk 1]
```diff
@@ -103,7 +103,7 @@ mod erc20 {
             let caller = self.env().caller();
             let allowance = self.allowance_of_or_zero(&from, &caller);
             if allowance < value {
-                return false
+                return false;
```
Comment (Robbepop) [inline examples/erc20/src/lib.rs]: I think this is missing a `rustfmt` run.

File: examples/erc20/src/lib.rs [hunk 2]
```diff
@@ -142,4 +142,109 @@ mod erc20 {
...
+        use super::*;
+        use ink_core::env;
+
+        /// The default constructor does its job.
+        #[test]
+        fn new_works() {
+            // Constructor works.
+            let _erc20 = Erc20::new(100);
+            // Transfer event triggered during initial contruction.
+            assert_eq!(1, env::test::recorded_events().count());
```
Comment (Robbepop) [inline examples/erc20/src/lib.rs]: Please no [Yoda conditions](https://en.wikipedia.org/wiki/Yoda_conditions).

File: examples/erc20/src/lib.rs [hunk 3]
```diff
@@ -142,4 +142,109 @@ mod erc20 {
...
+        }
+
+        /// The total supply was applied.
+        #[test]
+        fn total_supply_works() {
+            // Constructor works.
+            let erc20 = Erc20::new(100);
+            // Transfer event triggered during initial contruction.
+            assert_eq!(1, env::test::recorded_events().count());
```
Comment (Robbepop) [inline examples/erc20/src/lib.rs]: Conditions Yoda use not do please.

File: examples/erc20/src/lib.rs [hunk 4]
```diff
@@ -142,4 +142,109 @@ mod erc20 {
...
+            let accounts = env::test::default_accounts::<env::DefaultEnvTypes>()
+                .expect("Cannot get accounts");
+
+            assert_eq!(erc20.balance_of(accounts.bob), 0);
+            // Alice transfers 10 tokens to Bob.
+            assert_eq!(erc20.transfer(accounts.bob, 10), true);
+            // The second Transfer event takes place.
+            assert_eq!(2, env::test::recorded_events().count());
+            // Bob owns 10 tokens.
+            assert_eq!(erc20.balance_of(accounts.bob), 10);
```
Comment (Robbepop) [inline examples/erc20/src/lib.rs]: It might be nice to test for invalid transactions as well.

**Pull request merged**